include/jemalloc/internal/arena_inlines_b.h

```diff
@@ -40,23 +40,29 @@ arena_choose_maybe_huge(tsd_t *tsd, arena_t *arena, size_t size) {
 	return arena_choose(tsd, NULL);
 }
 
-JEMALLOC_ALWAYS_INLINE prof_tctx_t *
-arena_prof_tctx_get(tsdn_t *tsdn, const void *ptr, alloc_ctx_t *alloc_ctx) {
+JEMALLOC_ALWAYS_INLINE void
+arena_prof_info_get(tsdn_t *tsdn, const void *ptr, alloc_ctx_t *alloc_ctx,
+    prof_info_t *prof_info) {
 	cassert(config_prof);
 	assert(ptr != NULL);
+	assert(prof_info != NULL);
 
 	/* Static check. */
 	if (alloc_ctx == NULL) {
```

Comment:
It seems to me that this would end up making the logic more complicated, mainly due to:
- We need to define a local `alloc_ctx` variable if `alloc_ctx` is not given.
- We need to avoid the `iealloc()` call when `alloc_ctx` is given and the `slab` field is `true`, and we need to avoid calling `iealloc()` twice when `alloc_ctx` is not given but `extent_slab_get(extent)` turns out to be `false`.

The code would end up being something like:
```
        const extent_t *extent;
        alloc_ctx_t alloc_ctx_local;

        if (alloc_ctx == NULL) {
                extent = iealloc(tsdn, ptr);
                alloc_ctx_local.slab = extent_slab_get(extent);
                alloc_ctx = &alloc_ctx_local;
        } else if (unlikely(!alloc_ctx->slab)) {
                extent = iealloc(tsdn, ptr);
        }

        if (unlikely(!alloc_ctx->slab)) {
                large_prof_info_get(tsdn, extent, prof_info);
        } else {
                prof_info->prof_tctx = (prof_tctx_t *)(uintptr_t)1U;
        }
```
Any idea how to make it simpler?

Reply:
It looks we only need the `is_slab` part of the alloc_ctx.

Maybe like this:

```
extent = iealloc(tsdn, ptr);
bool is_slab = alloc_ctx != NULL ? alloc_ctx->slab : extent_slab_get(extent);
```

Reply:
It works other than that `iealloc()` is always called, and the call is unnecessary in the common case i.e. when `alloc_ctx != NULL && alloc_ctx->slab`.

How about:
```
const extent_t *extent = alloc_ctx != NULL && alloc_ctx->slab ? NULL : iealloc(tsdn, ptr);
/* extent is guaranteed to be non-NULL at this point if alloc_ctx != NULL. */
bool is_slab = alloc_ctx != NULL ? alloc_ctx->slab : extent_slab_get(extent);
```
(This code feels pretty tricky though. :) )

Reply:
Sorry in the comment I meant `if alloc_ctx == NULL`.

Reply:
Hmm this should work:

```
extent_t *e;
bool is_slab;
if (alloc_ctx != NULL) { 
    slab = alloc_ctx->slab;
    if (!slab) { e = iealloc(tsdn, ptr); }
} else {
    e = iealloc(tsdn, ptr);
    slab = extent_slab_get(extent);
}
if (unlikely(!slab)) {
    // compiler will warn if it find e is used uninitialized.
    large_prof_info_get(tsdn, e, prof_info);
} else {
    prof_info->prof_tctx = (prof_tctx_t *)(uintptr_t)1U;
}